### ISSUE TITLE
Infinite compiler loop

### ISSUE THREAD

**Issue by JoshuaOSHickman** [elm-lang/elm-compiler#1418]
So, I found myself writing 

```
iterateTillSame : (a -> a) -> a -> List a
iterateTillSame fn startingVal =
  let next = fn startingVal in
  if next == startingVal
    then [startingVal]
    else startingVal :: iterateTillSame next fn
```

There is a problem in this definition -- arguments are reversed in the recursive call. Oops! But if I try and compile this function, even with the type signature, it loops forever. I noticed there are warnings for infinite types (which is, I think, the issue here -- if I supply a monomorphic type signature it shows me the error), but this doesn't trigger the warnings, it just infinite loops.

**Comment by evancz** [comment]
I believe this is a duplicate of #1231, so I will close here. Thanks for the report!
